saleor/order/fetch.py

```diff
@@ -147,4 +153,61 @@ def fetch_draft_order_lines_info(
     ):
         voucher_info = fetch_voucher_info(voucher, order.voucher_code)
         apply_voucher_to_line(voucher_info, lines_info)
+        denormalized_voucher_info = _get_denormalized_voucher_info(lines_info, voucher)
+        _apply_denormalized_voucher_to_line_info(
+            lines_info, denormalized_voucher_info, order.voucher_code
+        )
     return lines_info
+
+
+def _get_denormalized_voucher_info(lines_info: list[EditableOrderLineInfo], voucher):
+    voucher_discounts = [
+        discount
+        for line_info in lines_info
+        for discount in line_info.discounts
+        if discount.voucher == voucher
+    ]
+    if not voucher_discounts:
+        return None
+
+    voucher_discount = voucher_discounts[0]
+    return VoucherDenormalizedInfo(
+        discount_value=voucher_discount.value,
+        discount_value_type=voucher_discount.value_type,
+        voucher_type=voucher.type,
+        reason=voucher_discount.reason,
+        name=voucher_discount.name,
+        apply_once_per_order=voucher.apply_once_per_order,
+        origin_line_id=voucher_discount.line_id,
+    )
+
+
+def _apply_denormalized_voucher_to_line_info(
+    lines_info: list[EditableOrderLineInfo],
+    denormalized_voucher_info: VoucherDenormalizedInfo | None,
+    voucher_code: str | None,
+):
+    if not denormalized_voucher_info:
+        return
+
+    # Denormalized vouchers of type SPECIFIC PRODUCT shouldn't be evaluated against
+    # the latest eligible product catalog so it should be applied to the same line
+    # it originates from
+    if denormalized_voucher_info.voucher_type == VoucherType.SPECIFIC_PRODUCT:
```

Comment:
Not sure if understand. You have a voucher for SPECIFIC_PRODUCTS, but you apply it only to single line. What if I have voucher for multiple products and draft order where multiple lines are applicable? This would cause, It will assign it only to a single line. What do I miss here?